### ISSUE TITLE
Track classification metrics (accuracy, precision, recall, F1, confusion matrix) alongside loss in evaluation epochs

### ISSUE THREAD

**Issue by LeoBuron** [es-ude/OnDeviceTraining#70]
## Motivation

Currently `trainingRun` computes loss per epoch for both training and evaluation, but accuracy is only available via a separate `evaluationEpochAccuracy` call after training finishes (as MnistExperiment does on its last line). This means:

- No per-epoch accuracy tracking during training
- The CSV log only contains `epoch, batch, train_loss, eval_loss` — no metric columns
- No way to monitor overfitting (accuracy diverging from loss trends) during training
- No precision, recall, or F1 tracking at all

### Eval-only metrics (no training accuracy/metrics)

Computing training metrics requires either a second forward pass per training sample (doubling compute cost per epoch) or reusing the output tensor from `calculateGradsSequential`. On MCUs where compute is precious, this overhead is not justified: training loss already signals whether the model is learning, and eval metrics are what matter for monitoring overfitting and generalization. Training metrics are dropped from scope.

### `numClasses` derived from tensor shapes — no new parameter

The current `evaluationBatchAccuracy` takes `numClasses` as an explicit parameter. But this is redundant: every label tensor has shape `[1, numClasses]`, so `calcNumberOfElementsByTensor(label)` returns `numClasses` directly. The output tensor from `inferenceWithLoss` is sized to match the label. Therefore `calcNumberOfElementsByTensor(stats->output)` gives us `numClasses` at the point of use, without threading a parameter through `trainingRun`, the eval functions, or storing it in the dataloader.

This avoids coupling the API to classification tasks and keeps function signatures unchanged where possible.

### Confusion matrix as the internal primitive

All classification metrics derive from the confusion matrix:

```
Confusion Matrix (numClasses × numClasses), cm[predicted][target] = count

  accuracy    = trace(cm) / total
  precision_c = cm[c][c] / row_sum(c)
  recall_c    = cm[c][c] / col_sum(c)
  f1_c        = 2 · precision_c · recall_c / (precision_c + recall_c)
  macro avg   = mean across all classes
```

Computing the CM costs the same as counting correct predictions: one argmax per sample, then `cm[predicted][target]++`. All metrics are derived from this single data structure.

### O(numClasses) memory for scalar metrics — confusion matrix is optional

The full confusion matrix is O(numClasses²). For 1000 classes (SmallImageNet), that's 8 MB (64-bit) — too expensive for MCUs and wasteful when only scalar metrics are needed.

However, all scalar metrics (accuracy, precision, recall, F1) are derivable from just **three arrays of size numClasses**:

| Data | What it stores | Space |
|------|---------------|-------|
| `tp[numClasses]` | True positives per class (CM diagonal) | O(n) |
| `predCount[numClasses]` | Predictions per class (CM row sums) | O(n) |
| `actualCount[numClasses]` | Actual labels per class (CM col sums) | O(n) |

Memory comparison:

| numClasses | 3 arrays (64-bit) | Full CM (64-bit) |
|------------|-------------------|------------------|
| 10 (MNIST) | 240 B | 800 B |
| 100 (CIFAR-100) | 2.4 KB | 80 KB |
| 1000 (ImageNet) | 24 KB | 8 MB |

Per-sample accumulation:
```c
size_t predicted = argmax(output, numClasses);
size_t target    = argmax(label, numClasses);
if (predicted == target) tp[predicted]++;
predCount[predicted]++;
actualCount[target]++;
// optionally: confusionMatrix[predicted * numClasses + target]++;
```

Therefore, scalar metric computation always uses the 3-array path. The full confusion matrix is only built when explicitly requested via `evaluationEpochWithReport`.

### Three public epoch functions — resource ladder

```
evaluationEpoch          →  loss only         → O(1) extra memory
evaluationEpochWithMetrics  →  scalar metrics → O(numClasses)
evaluationEpochWithReport   →  metrics + CM   → O(numClasses²)
```

Each level adds cost and data. The naming is self-documenting. Internally, `WithMetrics` and `WithReport` share private helpers — no code duplication. The loss-only `evaluationEpoch` is completely unchanged from the current implementation.

### Private helpers avoid code duplication

`evaluateBatchInternal` (private) does one loop per batch: extracts loss from `inferenceWithLoss`, runs argmax, accumulates into the 3 arrays and optionally the CM. Two thin public functions (`evaluationEpochWithMetrics` and `evaluationEpochWithReport`) call `evaluateEpochInternal` (private) which loops batches via the batch helper. The only difference: `WithReport` passes a pre-allocated CM pointer, `WithMetrics` passes NULL.

The loss-only path (`evaluationBatch` → `evaluationEpoch`) is untouched — no argmax, no arrays, no overhead.

### Not everything works with everything — by design

The `WithMetrics` and `WithReport` variants produce meaningful results only for classification tasks with one-hot labels. For regression (MSE), argmax over the output is nonsensical. `evaluationBatch` / `evaluationEpoch` remain the correct choice for non-classification tasks. A comment in `TrainingLoopApi.h` documents this constraint.

### Precision, recall, F1 are macro-averaged

Precision, recall, and F1 are per-class metrics. To produce a single scalar, we compute each metric per class and take the simple mean (macro average). Each class is weighted equally regardless of sample count.

Edge cases in metric derivation:
- Class with no predictions (row sum = 0): precision = 0
- Class with no actual samples (col sum = 0): recall = 0
- Both precision and recall = 0: F1 = 0

### New types

```c
typedef struct epochStats {
    float loss;
    float accuracy;
    float precision;    // macro-averaged
    float recall;       // macro-averaged
    float f1;           // macro-averaged
} epochStats_t;

typedef struct classificationReport {
    epochStats_t stats;
    size_t *confusionMatrix;   // numClasses × numClasses, row-major, heap-allocated
    size_t numClasses;
} classificationReport_t;
```

### Updated types

```c
typedef struct trainingRunResult {
    float finalTrainLoss;
    epochStats_t finalEvalStats;
} trainingRunResult_t;

typedef void (*epochCallbackFn_t)(size_t epoch, float trainLoss, epochStats_t evalStats);
```

`trainLoss` stays a bare `float` — there are no training metrics, so wrapping it in `epochStats_t` would be misleading.

### Function architecture

```
Unchanged (loss-only path):
  evaluationBatch  → float (avgLoss)
  evaluationEpoch  → float (avgLoss across batches)

Private (new, shared by metrics/report):
  evaluateBatchInternal(model, ..., batch, inferenceFn, tp, predCount, actualCount, cm, numClasses)
    → float (avgLoss), mutates arrays + optional CM

  evaluateEpochInternal(model, ..., dataLoader, inferenceFn, &outCM, &outNumClasses)
    → epochStats_t (derives metrics from 3 arrays)
    → optionally allocates CM, passes ownership to caller via outCM

  computeAccuracy(tp, actualCount, numClasses) → float
  computeMacroPrecision(tp, predCount, numClasses) → float
  computeMacroRecall(tp, actualCount, numClasses) → float
  computeMacroF1(tp, predCount, actualCount, numClasses) → float

Public (new):
  evaluationEpochWithMetrics → epochStats_t    (calls internal with outCM=NULL)
  evaluationEpochWithReport  → classificationReport_t (calls internal with outCM=&cm)
  freeClassificationReport   → void            (frees the heap-allocated CM)

Updated:
  trainingRun → trainingRunResult_t (param list unchanged, calls internal, frees CM per epoch)
```

### `evaluateBatchInternal` — the core shared function

```c
static float evaluateBatchInternal(layer_t **model, size_t modelSize,
                                    lossType_t lossType, batch_t *batch,
                                    inferenceWithLossFn_t inferenceFn,
                                    size_t *tp, size_t *predCount, size_t *actualCount,
                                    size_t *confusionMatrix, size_t numClasses) {
    float totalLoss = 0.0f;
    for (size_t i = 0; i < batch->size; i++) {
        inferenceStats_t *stats = inferenceFn(model, modelSize, batch->samples[i]->item,
                                               batch->samples[i]->label, lossType);
        totalLoss += stats->loss;

        float *outputData = (float *)stats->output->data;
        float *labelData  = (float *)batch->samples[i]->label->data;
        size_t predicted = argmax(outputData, numClasses);
        size_t target    = argmax(labelData, numClasses);

        if (predicted == target) tp[predicted]++;
        predCount[predicted]++;
        actualCount[target]++;

        if (confusionMatrix != NULL) {
            confusionMatrix[predicted * numClasses + target]++;
        }

        freeInferenceStats(stats);
    }
    return totalLoss / (float)batch->size;
}
```

### Data flow in `trainingRun`

```c
for (epoch) {
    float trainLoss = trainingEpochDefault(...);

    size_t *cm;
    size_t numClasses;
    epochStats_t evalStats = evaluateEpochInternal(..., &cm, &numClasses);
    freeReservedMemory(cm);

    if (callback) callback(epoch, trainLoss, evalStats);
    result.finalTrainLoss = trainLoss;
    result.finalEvalStats = evalStats;
}
```

The CM is built and freed every epoch inside the training loop. Scalar metrics survive in `epochStats_t`. For MNIST (10 classes), the CM is 800 bytes — trivial. For 1000 classes, the 3 arrays cost 24 KB; the transient CM costs 8 MB but is freed immediately.

**Note**: `trainingRun` currently builds the CM transiently to allow future use. If the 8 MB transient cost is unacceptable for 1000-class MCU training, `trainingRun` can be changed to pass `outCM=NULL` and compute metrics from the 3 arrays only. This is a one-line change and does not affect the public API.

### MnistExperiment changes

- `epochCallback` receives `epochStats_t evalStats`; writes all metrics to CSV
- CSV header: `epoch, train_loss, eval_loss, eval_accuracy, eval_precision, eval_recall, eval_f1`
- Remove standalone `evaluationEpochAccuracy` call — accuracy is now in `result.finalEvalStats.accuracy`
- Print all final metrics

### Removed

- `evaluationBatchAccuracy` — superseded by `evaluateBatchInternal`
- `evaluationEpochAccuracy` — superseded by `evaluationEpochWithMetrics`
- `inferenceFn_t` typedef — no longer used (metrics path uses `inferenceWithLossFn_t` via the shared helper)

## Files to Modify

- `src/userApi/training_loop/include/TrainingLoopApi.h` — new types (`epochStats_t`, `classificationReport_t`), updated types (`trainingRunResult_t`, `epochCallbackFn_t`), new declarations (`evaluationEpochWithMetrics`, `evaluationEpochWithReport`, `freeClassificationReport`), remove old (`evaluationBatchAccuracy`, `evaluationEpochAccuracy`, `inferenceFn_t`)
- `src/userApi/training_loop/TrainingLoopApi.c` — private helpers (`evaluateBatchInternal`, `evaluateEpochInternal`, metric derivation), new public functions, updated `trainingRun`, removed old functions
- `experiments/MnistExperiment.c` — updated callback, CSV columns, result access, remove standalone accuracy call
- `test/unit/userAPI/UnitTestTrainingLoopApi.c` — new tests, updated tests, removed old tests

## Test Plan

| Test | Status |
|------|--------|
| `testCalculateGradsSequential_MatchesPyTorch` | Unchanged |
| `testEvaluationBatch_ReturnsAverageLoss` | Unchanged |
| `testEvaluationEpoch_ReturnsAverageLossAcrossBatches` | Unchanged |
| `testEvaluationEpochWithMetrics_AllCorrect` | New — identity model, all correct: accuracy=1.0, precision=1.0, recall=1.0, F1=1.0 |
| `testEvaluationEpochWithMetrics_PartiallyCorrect` | New — designed for asymmetric CM, verify specific metric values |
| `testEvaluationEpochWithReport_ReturnsConfusionMatrix` | New — verify CM dimensions and cell values match expected |
| `testTrainingBatchDefault_ReturnsAverageLossAndAccumulatesGrads` | Unchanged |
| `testTrainingEpochDefault_DoesOptimizerStepPerBatch` | Unchanged |
| `testTrainingRun_ReturnsResult` | Updated: assert all `finalEvalStats` fields |
